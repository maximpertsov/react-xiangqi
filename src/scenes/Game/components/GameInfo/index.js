/** @jsx jsx */
import { jsx, css } from '@emotion/core';
import PropTypes from 'prop-types';

const GameInfo = ({ activePlayer, hasLegalMoves, userColor }) => {
  const userIsActive = () => {
    const { color } = activePlayer;
    return color === userColor;
  };

  const loadingUser = () => userColor === undefined;

  const getMessage = () => {
    if (loadingUser()) {
      const { color } = activePlayer;
      return `${color}'s turn`;
    }
    if (!hasLegalMoves) {
      // TODO: specify if won by stalemate or checkmate
      return userIsActive() ? 'You lose!' : 'You win!';
    }
    return userIsActive() ? 'Your turn' : 'Waiting for opponent';
  };

  return (
    <div
      className="GameInfo"
      css={css`
        color: #999;
      `}
    >
      <p>{ getMessage() }</p>
    </div>
  );
};

GameInfo.propTypes = {
  activePlayer: PropTypes.shape({
    color: PropTypes.string.isRequired,
    name: PropTypes.string,
  }),
  hasLegalMoves: PropTypes.bool.isRequired,
  userColor: PropTypes.string,
};

GameInfo.defaultProps = {
  activePlayer: undefined,
  userColor: undefined,
};

export default GameInfo;
