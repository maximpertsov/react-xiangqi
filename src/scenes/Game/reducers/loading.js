export const loading = (state = false, action) => {
  switch (action.type) {
  case 'toggle_loading':
    return action.loading;
  default:
    return state;
  }
};

export default {};
