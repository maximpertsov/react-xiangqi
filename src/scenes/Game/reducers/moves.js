import update from 'immutability-helper';
import sortedIndexBy from 'lodash/sortedIndexBy';

import XiangqiBoard, * as logic from 'services/logic';
import { Color } from 'services/logic/constants';
import { isRed } from 'services/logic/utils';

/* eslint-disable-next-line max-len */
const DEFAULT_FEN = 'rheakaehr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/9/RHEAKAEHR';

const initialMoves = [{
  id: 0,
  piece: undefined,
  fromPos: undefined,
  toPos: undefined,
  board: new XiangqiBoard({ fen: DEFAULT_FEN }),
  pending: false,
}];

const addMove = (state, { moveId, board, fromSlot, toSlot, pending }) => {
  const move = {
    id: moveId,
    fromPos: logic.getRankFile(fromSlot),
    toPos: logic.getRankFile(toSlot),
    piece: board.getPiece(fromSlot),
    board: board.move(fromSlot, toSlot),
    pending,
  };
  return update(state, { $push: [move] });
};

const cancelMoves = (state) => {
  const confirmedMoves = state.filter(({ pending }) => !pending);

  return confirmedMoves;
};

const confirmMoves = (state) => {
  return state.map((move) => ({ ...move, pending: false }));
};

// const setMove = (state, { origin: fromPos, destination: toPos }) => {
//   // TODO: get last move selector?
//   const { board } = state[state.length - 1];
//   const move = newMove({
//     piece: board.getPiece(fromPos, logic.RefType.RANK_FILE),
//     fromPos,
//     toPos,
//     board: board.move(fromPos, toPos, logic.RefType.RANK_FILE),
//     pending: false,
//   });
//   return update(state, { $push: [move] });
// };

// TODO: remove this?
// TODO: bulk increment nextMoveId counter?
const setMoves = () => initialMoves;

const moves = (state = initialMoves, action) => {
  switch (action.type) {
  case 'add_move':
    return addMove(state, action);
  case 'cancel_moves':
    return cancelMoves(state);
  case 'confirm_moves':
    return confirmMoves(state);
  case 'set_moves':
    return setMoves(state, action.moves);
  default:
    return state;
  }
};

export default moves;

/*******************/
/***  Selectors  ***/
/*******************/

const getMoveIndex = (state, moveId) => {
  const moveIndex = sortedIndexBy(state, { id: moveId }, 'id');
  if (state[moveIndex].id === moveId) return moveIndex;
  return -1;
};

export const getMoveCount = (state) => state.length - 1;

export const getLastMove = (state) => state[getMoveCount(state)];

export const getMoveById = (state, moveId) => {
  const moveIndex = getMoveIndex(state, moveId);
  return state[moveIndex];
};

export const getPreviousMove = (state, moveId) => {
  const moveIndex = getMoveIndex(state, moveId);
  return state[Math.min(moveIndex - 1, 0)];
};

export const getNextMove = (state, moveId) => {
  const moveCount = getMoveCount(state);
  const moveIndex = getMoveIndex(state, moveId);
  return state[Math.max(moveIndex + 1, moveCount)];
};

export const getNextMoveColor = (state) => {
  if (getMoveCount(state) === 0) return Color.RED;

  const { piece } = getLastMove(state);
  return isRed(piece) ? Color.BLACK : Color.RED;
};
