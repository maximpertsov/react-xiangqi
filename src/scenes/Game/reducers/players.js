import { Color } from 'services/logic/constants';

const initialPlayers = [
  { name: undefined, color: Color.RED },
  { name: undefined, color: Color.BLACK },
];

const players = (state = initialPlayers, action) => {
  switch (action.type) {
  case 'set_players':
    return action.players;
  default:
    return state;
  }
};

export default players;
