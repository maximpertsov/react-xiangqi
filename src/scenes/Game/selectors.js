import sortedIndexBy from 'lodash/sortedIndexBy';

import { Color } from 'services/logic/constants';
import { isRed } from 'services/logic/utils';

/***************/
/***  Moves  ***/
/***************/

export const getFirstMove = ({ moves }) =>
  moves[0];

export const getLastMove = ({ moves }) =>
  moves[moves.length - 1];

export const getMoveIndex = ({ moves }, moveId ) => {
  const moveIndex = sortedIndexBy(moves, { id: moveId }, 'id');
  if (moves[moveIndex].id === moveId) return moveIndex;
  return -1;
};

// TODO: duplicate logic in actions.js
export const getSelectedMove = ({ moves, selectedMoveId }) =>
  moves[getMoveIndex({ moves }, selectedMoveId)];

export const getNextMoveColor = ({ moves }) => {
  if (moves.length === 0) return Color.RED;

  const { piece } = getLastMove({ moves });
  return isRed(piece) ? Color.BLACK : Color.RED;
};

/*****************/
/***  Players  ***/
/*****************/

const lookupPlayer = (players, key, value) =>
  players.find((p) => p[key] === value);

export const getNextMovePlayer = ({ players, moves }) =>
  lookupPlayer(players, 'color', getNextMoveColor({ moves }));

const getUserPlayer = ({ players }, { username }) =>
  lookupPlayer(players, 'name', username);

export const getRedPlayer = ({ players }) =>
  lookupPlayer(players, 'color', Color.RED);

export const getBlackPlayer = ({ players }) =>
  lookupPlayer(players, 'color', Color.BLACK);

export const getUserColor = ({ players }, { username }) => {
  try {
    return getUserPlayer({ players }, { username }).color;
  } catch (e) {
    if (e instanceof TypeError) return undefined;
    throw e;
  }
};

export const getOtherPlayer = ({ players }, { gameSlug, username }) => {
  if (gameSlug === undefined) getBlackPlayer({ players });
  return players.find((p) => p.name !== username);
};

// TODO: add a state that allows players to flip their original orientation
export const getInitialUserOrientation = ({ players }, { username }) =>
  getUserColor({ players }, { username }) === Color.BLACK;

// TODO: move to layout class that displays board and players
export const getCurrentPlayer = ({ players }, { gameSlug, username }) => {
  if (gameSlug === undefined) getRedPlayer({ players });
  return getUserPlayer({ players }, { username });
};

/********************/
/***  Game Logic  ***/
/********************/

// TODO break up function
export const getLegalMoves = (
  { moves, players, selectedMoveId },
  { gameSlug, username },
) => {
  const nextMoveColor = getNextMoveColor({ moves });
  const userColor = getUserColor({ players }, { username });
  const { board } = getSelectedMove({ moves, selectedMoveId });
  const currentUserOnly = gameSlug !== undefined;
  const lastMoveId = getLastMove({ moves }).id;

  return board
    .legalMoves()
    .map((toSlots, fromSlot) => {
      if (selectedMoveId !== lastMoveId) return [];
      if (!board.isColor(nextMoveColor, fromSlot)) return [];
      if (currentUserOnly && !board.isColor(userColor, fromSlot)) return [];
      return toSlots;
    });
};

// TODO: this isn't really a selector -- use refactor with mapStateToProps
export const hasLegalMoves = (state) => {
  const { board } = getLastMove(state);
  return board.hasLegalMoves(getNextMoveColor(state));
};

export default {};
