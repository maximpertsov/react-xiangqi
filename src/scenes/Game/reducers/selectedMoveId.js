import { getMoveCount, getMoveIndex } from '../selectors';

const setSelectedMoveIndex = (moves, moveIndex) => {
  const { id } = moves[moveIndex];
  return id;
};

const setPreviousMove = (state, moves) => {
  const moveIndex = getMoveIndex({ moves }, state);
  return setSelectedMoveIndex(moves, Math.max(moveIndex - 1, 0));
};

const setNextMove = (state, moves) => {
  const moveIndex = getMoveIndex({ moves }, state);
  const moveCount = getMoveCount({ moves });
  return setSelectedMoveIndex(moves, Math.min(moveIndex + 1, moveCount));
};

const selectLastMove = (state, moves) => {
  const moveCount = getMoveCount({ moves });
  return setSelectedMoveIndex(moves, moveCount);
};

const selectedMoveId = (state = 0, action) => {
  switch (action.type) {
  case 'select_move':
    return action.moveId;
  case 'select_previous_move':
    return setPreviousMove(state, action.moves);
  case 'select_next_move':
    return setNextMove(state, action.moves);
  case 'select_last_move':
    return selectLastMove(state, action.moves);
  default:
    return state;
  }
};


export default selectedMoveId;
