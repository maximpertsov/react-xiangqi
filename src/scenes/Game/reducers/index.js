import { combineReducers } from 'redux';
import loading from 'scenes/Game/reducers/loading';
import moves, * as fromMoves from 'scenes/Game/reducers/moves';
import players from 'scenes/Game/reducers/players';
import selectedMoveId from 'scenes/Game/reducers/selectedMoveId';

const game = combineReducers({ loading, moves, players, selectedMoveId });

export default game;

/***************/
/***  Moves  ***/
/***************/

export const getMoveCount = ({ moves }) => fromMoves.getMoveCount(moves);

export const getLastMove = ({ moves }) => fromMoves.getLastMove(moves);

export const getSelectedMove = ({ moves, selectedMoveId }) =>
  fromMoves.getMoveById(moves, selectedMoveId);

export const getPreviousMove = ({ moves, selectedMoveId }) =>
  fromMoves.getPreviousMove(moves, selectedMoveId);

export const getNextMove = ({ moves, selectedMoveId }) =>
  fromMoves.getNextMove(moves, selectedMoveId);

export const getNextMoveColor = ({ moves }) =>
  fromMoves.getNextMoveColor(moves);
