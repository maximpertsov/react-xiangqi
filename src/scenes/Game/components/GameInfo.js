import React, { useCallback, useEffect } from 'react';
import styled from '@emotion/styled';
import { createSelector } from 'reselect';
import { useSelector } from 'react-redux';
import { getNextMovePlayer, getCurrentPlayer } from 'reducers';

import isEqual from 'lodash/isEqual';

import { Team } from 'services/logic/constants';

const Wrapper = styled.div`
  align-items: center;
  color: #999;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
`;

// eslint-disable-next-line complexity
const getGameOverMessage = ({ player2, score2, player1, score1 }) => {
  if (score1 === 1) return `${player1.name} wins!`;
  if (score2 === 1) return `${player2.name} wins!`;
  if (score1 === 0.5 && score2 === 0.5) return 'Draw!';
};

const getGameInProgressMessage = ({
  currentPlayer,
  nextMovePlayer,
  username,
}) => {
  if (!username && nextMovePlayer.team === Team.RED) {
    return 'Your turn';
  }
  if (isEqual(nextMovePlayer, currentPlayer)) {
    return 'Your turn';
  }

  return 'Waiting for opponent';
};

const mapStateToProps = createSelector(
  state => state.player2,
  state => state.score2,
  state => state.player1,
  state => state.score1,
  state => state.username,
  state => getCurrentPlayer(state),
  state => getNextMovePlayer(state),

  (
    player2,
    score2,
    player1,
    score1,
    username,
    currentPlayer,
    nextMovePlayer,
  ) => ({
    gameOverMessage: getGameOverMessage({
      player2,
      score2,
      player1,
      score1,
    }),
    gameInProgressMessage: getGameInProgressMessage({
      currentPlayer,
      nextMovePlayer,
      username,
    }),
  }),
);

const GameInfo = () => {
  const { gameOverMessage, gameInProgressMessage } = useSelector(
    mapStateToProps,
    isEqual,
  );

  const notify = useCallback(() => {
    if (window.document.hasFocus()) return;

    new window.Notification(gameOverMessage || gameInProgressMessage);
  }, [gameInProgressMessage, gameOverMessage]);

  useEffect(() => {
    window.document.title = gameOverMessage || gameInProgressMessage;

    if (!('Notification' in window)) return;

    switch (window.Notification.permission) {
      case 'denied':
        break;
      case 'granted':
        notify();
        break;
      default:
        window.Notification.requestPermission(permission => {
          if (permission !== 'granted') return;

          notify();
        });
    }
  }, [gameInProgressMessage, gameOverMessage, notify]);

  return (
    <Wrapper className="GameInfo">
      <p>{gameOverMessage || gameInProgressMessage}</p>
    </Wrapper>
  );
};

export default GameInfo;
