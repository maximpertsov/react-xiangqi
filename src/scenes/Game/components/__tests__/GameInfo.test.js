import React from 'react';
import * as redux from 'react-redux';
import { shallow } from 'enzyme';
import values from 'lodash/values';
import * as selectors from 'reducers/selectors';

import GameInfo from 'scenes/Game/components/GameInfo';

const defaultState = {
  player1: { name: 'alice' },
  player2: { name: 'bob' },
};

const getShallowWrappedComponent = store =>
  shallow(
    <redux.Provider store={store}>
      <GameInfo />
    </redux.Provider>,
  )
    .dive()
    .dive();

describe('GameInfo', () => {
  let store;
  let wrapper;
  let spys = {};

  beforeEach(() => {
    spys.useSelector = jest.spyOn(redux, 'useSelector');
    spys.useSelector.mockImplementation(callback => callback(store.getState()));

    spys.getCurrentPlayer = jest.spyOn(selectors, 'getCurrentPlayer');
    spys.getCurrentPlayer.mockReturnValue({ name: 'alice' });

    wrapper = getShallowWrappedComponent(store);
  });

  afterEach(() => {
    store.clearActions();
    values(spys).forEach(spy => spy.mockRestore());
  });

  describe('Your turn', () => {
    beforeAll(() => {
      store = mockStore(defaultState);

      spys.getNextMovePlayer = jest.spyOn(selectors, 'getNextMovePlayer');
      spys.getNextMovePlayer.mockReturnValue({ name: 'alice' });
      console.log("your turn mock set");
    });

    test('snapshot', () => {
      console.log("your turn test");
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Waiting for opponent', () => {
    beforeAll(() => {
      store = mockStore(defaultState);

      spys.getNextMovePlayer = jest.spyOn(selectors, 'getNextMovePlayer');
      spys.getNextMovePlayer.mockReturnValue({ name: 'bob' });
      console.log("waiting mock");
    });

    test('snapshot', () => {
      console.log("waiting test");
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Player 1 (Alice) wins', () => {
    beforeAll(() => {
      store = mockStore({
        ...defaultState,
        score1: 1,
      });
      console.log("p1 wins mock");
    });

    test('snapshot', () => {
      console.log("p1 wins test");
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Player 2 (Bob) wins', () => {
    beforeAll(() => {
      store = mockStore({
        ...defaultState,
        score2: 1,
      });
      console.log("p2 wins mock");
    });

    test('snapshot', () => {
      console.log("p2 wins test");
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Draw', () => {
    beforeAll(() => {
      store = mockStore({
        ...defaultState,
        score1: 0.5,
        score2: 0.5,
      });
      console.log("draw mock");
    });

    test('snapshot', () => {
      console.log("draw test");
      expect(wrapper).toMatchSnapshot();
    });
  });
});
