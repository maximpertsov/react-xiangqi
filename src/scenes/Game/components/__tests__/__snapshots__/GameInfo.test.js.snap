// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameInfo Draw snapshot 1`] = `
<Styled(div)
  className="GameInfo"
>
  <p>
    Draw!
  </p>
</Styled(div)>
`;

exports[`GameInfo Player 1 (Alice) wins snapshot 1`] = `
<Styled(div)
  className="GameInfo"
>
  <p>
    alice wins!
  </p>
</Styled(div)>
`;

exports[`GameInfo Player 2 (Bob) wins snapshot 1`] = `
<Styled(div)
  className="GameInfo"
>
  <p>
    bob wins!
  </p>
</Styled(div)>
`;

exports[`GameInfo Waiting for opponent snapshot 1`] = `
<Styled(div)
  className="GameInfo"
>
  <p>
    Waiting for opponent
  </p>
</Styled(div)>
`;

exports[`GameInfo Your turn snapshot 1`] = `
<Styled(div)
  className="GameInfo"
>
  <p>
    Your turn
  </p>
</Styled(div)>
`;
